src/line_extractor_igvc/test/DBSCAN-test.cpp

```diff
@@ -0,0 +1,438 @@
+        p.x = x;
+        p.y = y2;
+        pcl.push_back(p);
+    }
+
+    pcl::PointCloud<pcl::PointXYZ>::Ptr pclPtr = pcl.makeShared();
+
+    std::vector<pcl::PointCloud<pcl::PointXYZ>> clusters =
+    dbscan.findClusters(pclPtr);
+    EXPECT_EQ(2, clusters.size());
+    EXPECT_EQ(firstLineLength, clusters[0].size());
+    EXPECT_EQ(secondLineLength, clusters[1].size());
+}
+
+TEST(DBSCAN, TestClusterBorder) {
+    int min_neighbours = 1;
+    int radius         = 6;
+    DBSCAN dbscan(min_neighbours, radius);
+
+    pcl::PointCloud<pcl::PointXYZ> pcl;
+
+    //    first line
+    float y1            = 3.000001;
+    int firstLineLength = 3;
+    int xinit           = -10;
+    for (int x = xinit; x < xinit + firstLineLength; x++) {
+        pcl::PointXYZ p;
+        p.x = x;
+        p.y = y1;
+        pcl.push_back(p);
+    }
+    // second line
+    float y2             = -3;
+    int secondLineLength = 3;
+    for (int x = xinit; x < xinit + secondLineLength; x++) {
+        pcl::PointXYZ p;
+        p.x = x;
+        p.y = y2;
+        pcl.push_back(p);
+    }
+
+    pcl::PointCloud<pcl::PointXYZ>::Ptr pclPtr = pcl.makeShared();
+
+    vector<pcl::PointCloud<pcl::PointXYZ>> clusters =
+    dbscan.findClusters(pclPtr);
+    EXPECT_EQ(2, clusters.size());
+    EXPECT_EQ(firstLineLength, clusters[0].size());
+    EXPECT_EQ(secondLineLength, clusters[1].size());
+}
+
+TEST(DBSCAN, TestClusterTwoPolynomialLines) {
+    int min_neighbours = 1;
+    int radius         = 3;
+    DBSCAN dbscan(min_neighbours, radius);
+
+    pcl::PointCloud<pcl::PointXYZ> pcl;
+
+    //    first line
+    int y1              = 10;
+    int firstLineLength = 40;
+    int xinit           = -20;
+    float a             = 0.002;
+    for (int x = xinit; x < xinit + firstLineLength; x++) {
+        pcl::PointXYZ p;
+        p.x = x;
+        p.y = a * pow(x, 3) + y1;
+        pcl.push_back(p);
+    }
+    // second line
+    int y2               = -10;
+    int secondLineLength = 40;
+    for (int x = xinit; x < xinit + secondLineLength; x++) {
+        pcl::PointXYZ p;
+        p.x = x;
+        p.y = a * pow(x, 3) + y2;
+        pcl.push_back(p);
+    }
+
+    pcl::PointCloud<pcl::PointXYZ>::Ptr pclPtr = pcl.makeShared();
+
+    vector<pcl::PointCloud<pcl::PointXYZ>> clusters =
+    dbscan.findClusters(pclPtr);
+    EXPECT_EQ(2, clusters.size());
+    EXPECT_EQ(firstLineLength, clusters[0].size());
+    EXPECT_EQ(secondLineLength, clusters[1].size());
+}
+
+TEST(DBSCAN, TestClusterOnePolynomialLinesWithNoise) {
+    int min_neighbours = 1;
+    int radius         = 80;
+    DBSCAN dbscan(min_neighbours, radius);
+
+    pcl::PointCloud<pcl::PointXYZ> pcl;
+
+    int y1   = 1000;
+    float m1 = 7;
+    float m2 = -0.7;
+    float m3 = 0.007;
+
+    int num_points = 100;
+
+    double truemin, truemax;
+
+    for (float x = 0; x < num_points; x++) {
+        float true_y = m1 * x + m2 * pow(x, 2) + m3 * pow(x, 3) + y1;
+        float true_x = x;
```

Comment:
It looks like we construct lines like this several times across multiple test files. It might be helpful to create a `TestUtils.h` file that holds the functions to do these sorts of things, which should make it easier to document and modify. Alternatively, we could just add these functions to the top of the test files where we're using them (which would make more sense if each set of tests uses a different version of the function). Totally up to your discretion, whatever you think would be cleaner!

Reply:
Done